### PR TITLE
[Harness] Backport all needed commits to allow to use tcp tunnels on iOS 14

### PR DESCRIPTION
**Pull request by mandel-macaque**
Most if not all of the iOS 14 device tests fail due to the new security changes form apple: https://developer.apple.com/videos/play/wwdc2020/10110/

In order to workaround this problem, we can use the tunnel bore code that is present in main. This PR contains all the different changes that were added in main up to the point of the tunnel bore generation. All the commits have been cherry-picked and some of the merge issues have been solved.

WARNING: DO NOT SQUASH THIS PR.

### CODE REVIEW

File: tests/xharness/Jenkins/Jenkins.cs [hunk 1]
```diff
@@ -554,7 +561,13 @@ async Task<IEnumerable<TestTask>> CreateRunSimulatorTasksAsync ()
...
 
 			var testVariations = CreateTestVariations (runSimulatorTasks, (buildTask, test, candidates) =>
-				new RunSimulatorTask (simulators, buildTask, processManager, candidates?.Cast<SimulatorDevice> () ?? test.Candidates)).ToList ();
+				new RunSimulatorTask (
+					jenkins: this,
+					simulators: simulators,
+					buildTask: buildTask,
+					processManager: processManager,
+					tunnelBore: tunnelBore,
+					candidates: candidates?.Cast<SimulatorDevice> () ?? test.Candidates)).ToList ();
```
Comment (whitneyschmidt) [inline tests/xharness/Jenkins/Jenkins.cs:570]: A little confused about `candidates` here. Why do we need to switch between these two different values?

File: tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/IResourceManager.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using System.Collections.Generic;
+using Microsoft.DotNet.XHarness.iOS.Shared.Hardware;
+
+namespace Microsoft.DotNet.XHarness.iOS.Shared.Tasks {
+
+	/// <summary>
+	/// Managed the resorces that can be used by the different tests tasks so that when ran async they do not step
+	/// on eachother.
```
Comment (whitneyschmidt) [inline tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/IResourceManager.cs:8]:
```suggestion
	/// on each other.
```

File: tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/BuildProject.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+using System.Threading.Tasks;
+using System.Xml;
+using Microsoft.DotNet.XHarness.iOS.Shared.Execution;
+using Microsoft.DotNet.XHarness.iOS.Shared.Logging;
+using Microsoft.DotNet.XHarness.iOS.Shared.Utilities;
+
+namespace Microsoft.DotNet.XHarness.iOS.Shared.Tasks {
+	public class BuildProject : BuildTool {
+		public IResourceManager ResourceManager { get; set; }
+		public IEnvManager EnviromentManager { get; set; }
```
Comment (whitneyschmidt) [inline tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/BuildProject.cs:14]:
```suggestion
		public IEnvManager EnvironmentManager { get; set; }
```

File: tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/BuildProject.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		string msbuildPath;
+
+		public string SolutionPath { get; set; }
+
+		public BuildProject (string msbuildPath, IProcessManager processManager, IResourceManager resourceManager, IEventLogger eventLogger, IEnvManager envManager) : base (processManager)
+		{
+			this.msbuildPath = msbuildPath ?? throw new ArgumentNullException (nameof (msbuildPath));
+			ResourceManager = resourceManager ?? throw new ArgumentNullException (nameof (resourceManager));
+			EventLogger = eventLogger ?? throw new ArgumentNullException (nameof (eventLogger));
+			EnviromentManager = envManager ?? throw new ArgumentNullException (nameof (envManager));
```
Comment (whitneyschmidt) [inline tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/BuildProject.cs:25]:
```suggestion
			EnvironmentManager = envManager ?? throw new ArgumentNullException (nameof (envManager));
```

File: tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/BuildProject.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+					args.Add ((useXIBuild && !isSolution ? "/" : "") + "restore"); // diff param depending on the tool
+					args.Add (projectPath);
+					if (useXIBuild && !isSolution)
+						args.Add ("/verbosity:detailed");
+					else {
+						args.Add ("-verbosity");
+						args.Add ("detailed");
+					}
+					nuget.StartInfo.Arguments = StringUtils.FormatArguments (args);
+					EnviromentManager.SetEnvironmentVariables (nuget);
```
Comment (whitneyschmidt) [inline tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/BuildProject.cs:61]:
```suggestion
					EnvironmentManager.SetEnvironmentVariables (nuget);
```

File: tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/MSBuild.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+																									ILog mainLog)
+		{
+			BuildLog = buildLog;
+			(TestExecutingResult ExecutionResult, string KnownFailure) result = (TestExecutingResult.NotStarted, (string) null);
+			await RestoreNugetsAsync (buildLog, resource, useXIBuild: true);
+
+			using (var xbuild = new Process ()) {
+				xbuild.StartInfo.FileName = msbuildPath;
+				xbuild.StartInfo.Arguments = StringUtils.FormatArguments (GetToolArguments (projectPlatform, projectConfiguration, projectFile, buildLog));
+				EnviromentManager.SetEnvironmentVariables (xbuild);
```
Comment (whitneyschmidt) [inline tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/MSBuild.cs:57]:
```suggestion
				EnvironmentManager.SetEnvironmentVariables (xbuild);

```

File: tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/MSBuild.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+				args.Add ("--");
+				args.Add ("/verbosity:diagnostic");
+				if (project_platform != null)
+					args.Add ($"/p:Platform={project_platform}");
+				if (project_configuration != null)
+					args.Add ($"/p:Configuration={project_configuration}");
+				args.Add (project_file);
+				args.Add ("/t:Clean");
+				xbuild.StartInfo.Arguments = StringUtils.FormatArguments (args);
+				EnviromentManager.SetEnvironmentVariables (xbuild);
```
Comment (whitneyschmidt) [inline tests/xharness/Microsoft.DotNet.XHarness.iOS.Shared/Tasks/MSBuild.cs:94]:
```suggestion
				EnvironmentManager.SetEnvironmentVariables (xbuild);

```

**Comment by dalexsoto** [comment]
Since this is a backport I guess we can keep the typos or fix in main and then backport again in a separated PR

**Comment by whitneyschmidt** [comment]
> Since this is a backport I guess we can keep the typos or fix in main and then backport again in a separated PR

Hmmm...probably not worth backporting for a handful of small typos in our internal tooling 🤦‍♀️

**Comment by mandel-macaque** [comment]
@dalexsoto @whitneyschmidt I'm not fixing the typos because this are cherry-picks, but I'll add them to main.

**Pull request merged**